[SPARK-7076][SPARK-7077][SPARK-7080][SQL] Use managed memory for aggregations

JoshRosen opened the pull request:
This patch adds managed-memory-based aggregation to Spark SQL / DataFrames. Instead of working with Java objects, this new aggregation path uses `sun.misc.Unsafe` to manipulate raw memory.  This reduces the memory footprint for aggregations, resulting in fewer spills, OutOfMemoryErrors, and garbage collection pauses.  As a result, this allows for higher memory utilization.  It can also result in better cache locality since objects will be stored closer together in memory.

This feature can be eanbled by setting `spark.sql.unsafe.enabled=true`.  For now, this feature is only supported when codegen is enabled and only supports aggregations for which the grouping columns are primitive numeric types or strings and aggregated values are numeric.
### Managing memory with sun.misc.Unsafe

This patch supports both on- and off-heap managed memory.
- In on-heap mode, memory addresses are identified by the combination of a base Object and an offset within that object.
- In off-heap mode, memory is addressed directly with 64-bit long addresses.

To support both modes, functions that manipulate memory accept both `baseObject` and `baseOffset` fields.  In off-heap mode, we simply pass `null` as `baseObject`.

We allocate memory in large chunks, so memory fragmentation and allocation speed are not significant bottlenecks.

By default, we use on-heap mode.  To enable off-heap mode, set `spark.unsafe.offHeap=true`.

To track allocated memory, this patch extends `SparkEnv` with an `ExecutorMemoryManager` and supplies each `TaskContext` with a `TaskMemoryManager`.  These classes work together to track allocations and detect memory leaks.
### Compact tuple format

This patch introduces `UnsafeRow`, a compact row layout.  In this format, each tuple has three parts: a null bit set, fixed length values, and variable-length values:

![image](https://cloud.githubusercontent.com/assets/50748/7328538/2fdb65ce-ea8b-11e4-9743-6c0f02bb7d1f.png)
- Rows are always 8-byte word aligned (so their sizes will always be a multiple of 8 bytes)
- The bit set is used for null tracking:
  - Position _i_ is set if and only if field _i_ is null
  - The bit set is aligned to an 8-byte word boundary.
- Every field appears as an 8-byte word in the fixed-length values part:
  - If a field is null, we zero out the values.
  - If a field is variable-length, the word stores a relative offset (w.r.t. the base of the tuple) that points to the beginning of the field's data in the variable-length part.
- Each variable-length data type can have its own encoding:
  - For strings, the first word stores the length of the string and is followed by UTF-8 encoded bytes.  If necessary, the end of the string is padded with empty bytes in order to ensure word-alignment.

For example, a tuple that consists 3 fields of type (int, string, string), with value (null, “data”, “bricks”) would look like this:

![image](https://cloud.githubusercontent.com/assets/50748/7328526/1e21959c-ea8b-11e4-9a28-a4350fe4a7b5.png)

This format allows us to compare tuples for equality by directly comparing their raw bytes.  This also enables fast hashing of tuples.
### Hash map for performing aggregations

This patch introduces `UnsafeFixedWidthAggregationMap`, a hash map for performing aggregations where the aggregation result columns are fixed-with.  This map's keys and values are `Row` objects. `UnsafeFixedWidthAggregationMap` is implemented on top of `BytesToBytesMap`, an append-only map which supports byte-array keys and values.

`BytesToBytesMap` stores pointers to key and value tuples.  For each record with a new key, we copy the key and create the aggregation value buffer for that key and put them in a buffer. The hash table then simply stores pointers to the key and value. For each record with an existing key, we simply run the aggregation function to update the values in place.

This map is implemented using open hashing with triangular sequence probing.  Each entry stores two words in a long array: the first word stores the address of the key and the second word stores the relative offset from the key tuple to the value tuple, as well as the key's 32-bit hashcode.  By storing the full hashcode, we reduce the number of equality checks that need to be performed to handle position collisions ()since the chance of hashcode collision is much lower than position collision).

`UnsafeFixedWidthAggregationMap` allows regular Spark SQL `Row` objects to be used when probing the map.  Internally, it encodes these rows into `UnsafeRow` format using `UnsafeRowConverter`.  This conversion has a small overhead that can be eliminated in the future once we use UnsafeRows in other operators.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/apache/spark/5725)

<!-- Reviewable:end -->

JoshRosen committed:
Initial import of code from Databricks unsafe utils repo.

JoshRosen committed:
Begin merging the UTF8String implementations.

JoshRosen committed:
Play around with Unsafe implementations of more string methods.

JoshRosen committed:
Add skeleton for Row implementation.

JoshRosen committed:
Add skeleton for GeneratedAggregate integration.

This typechecks properly and sketches how I'm intending to use row pointers
and the hashmap. This has been a useful exercise for figuring out whether my
interfaces will be sufficient.

JoshRosen committed:
Add reminder to free memory on iterator completion

JoshRosen committed:
Start prototyping Java Row -> UnsafeRow converters

JoshRosen committed:
Sketch how the converters will be used in UnsafeGeneratedAggregate

JoshRosen committed:
First version that passes some aggregation tests:

I commented out a number of tests where we do not support the required
data types; this is only a short-term hack until I extend the planner
to understand when UnsafeGeneratedAggregate can be used.

JoshRosen committed:
Some clarification of the BytesToBytesMap.lookup() / set() contract.

JoshRosen committed:
Simplify address + length calculation in Location.

JoshRosen committed:
Now that the store*() contract has been stregthened, we can remove an extra lookup

JoshRosen committed:
Add finalizer as "cleanup method of last resort"

JoshRosen committed:
Reuse UnsafeRow pointer in UnsafeRowConverter

JoshRosen committed:
Optimize for fact that get() is only called on String columns.

JoshRosen committed:
Free buffer once iterator has been fully consumed.

JoshRosen committed:
Add feature flag to guard UnsafeGeneratedAggregate

JoshRosen committed:
Fix bug in UnsafeMemoryAllocator.free():

The `if` check here was backwards, which prevented any memory from
being freed.

JoshRosen committed:
Optimizations related to zeroing out memory:

- Do not zero out all allocated memory; the zeroing isn't free and in many
  cases it isn't necessary.
- There are some cases where we do want to clear the memory, such as in BitSet.

  It shouldn't be the BitSet object's responsibility to zero out the memory
  block passed to it (since maybe we're passing some memory created by someone
  else and want to interpret it as a bitset). To make the caller's life easier,
  though, I added a MemoryBlock.zero() method for clearing the block.

- In UnsafeGeneratedAggregate, use Arrays.fill to clear the re-used temporary
  row buffer, since this is likely to be much faster than Unsafe.setMemory;
  see http://psy-lob-saw.blogspot.com/2015/04/on-arraysfill-intrinsics-superword-and.html
  for more details.

JoshRosen committed:
Use UNSAFE allocator in GeneratedAggregate (TODO: make this configurable)

JoshRosen committed:
Update to implement new Row methods added upstream

JoshRosen committed:
Extract aggregation map into its own class.

This makes the code much easier to understand and
will allow me to implement unsafe versions of both
GeneratedAggregate and the regular Aggregate operator.

JoshRosen committed:
Comment update (bumping to refresh GitHub cache...)

JoshRosen committed:
Add basic sanity test for UnsafeFixedWidthAggregationMap

JoshRosen committed:
Merge Unsafe code into the regular GeneratedAggregate, guarded by a
configuration flag; integrate planner support and re-enable all tests.

JoshRosen committed:
Address a number of minor code review comments.

JoshRosen committed:
Use ints intead of longs for indexing.

JoshRosen committed:
Bump up default hash map size

JoshRosen committed:
Enable optional performance metrics for hash map.

JoshRosen committed:
Fix bug in murmur hash implementation.

JoshRosen committed:
More long -> int conversion.

JoshRosen committed:
Remove debug logging

JoshRosen committed:
Fix a bunch of FindBugs and IntelliJ inspections

JoshRosen committed:
More warning cleanup

JoshRosen committed:
Import and comment cleanup

JoshRosen committed:
More mod replacement

JoshRosen committed:
Some Javadoc cleanup

JoshRosen committed:
Cleanups to string handling code

Changes to UTF8String will be deferred for a followup, since we should
benchmark them first.

JoshRosen committed:
Lots of TODO and doc cleanup.

JoshRosen committed:
Null handling improvements in UnsafeRow.

JoshRosen committed:
Add configuration options for heap vs. offheap

JoshRosen committed:
Add missing pom.xml

SparkQA commented:
[Test build #31059 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31059/consoleFull) for   PR 5725 at commit [`cde4132`](https://github.com/apache/spark/commit/cde413249c0083f175843b3d25023a76948692b2).

SparkQA commented:
[Test build #31059 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31059/consoleFull) for   PR 5725 at commit [`cde4132`](https://github.com/apache/spark/commit/cde413249c0083f175843b3d25023a76948692b2).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public final class UnsafeFixedWidthAggregationMap `
  * `  public static class MapEntry `
  * `public final class UnsafeRow implements MutableRow `
  * `class UnsafeRowConverter(fieldTypes: Array[DataType]) `
  * `public final class PlatformDependent `
  * `public class ByteArrayMethods `
  * `public final class LongArray `
  * `public final class BitSet `
  * `public final class BitSetMethods `
  * `public final class Murmur3_x86_32 `
  * `public final class BytesToBytesMap `
  * `  public final class Location `
  * `  class Doubling implements HashMapGrowthStrategy `
  * `public class HeapMemoryAllocator implements MemoryAllocator `
  * `public class MemoryBlock extends MemoryLocation `
  * `public class MemoryLocation `
  * `public class UnsafeMemoryAllocator implements MemoryAllocator `

 * This patch **adds the following new dependencies:**
   * `spark-unsafe_2.10-1.4.0-SNAPSHOT.jar`
   * `tachyon-0.6.4.jar`
   * `tachyon-client-0.6.4.jar`

 * This patch **removes the following dependencies:**
   * `tachyon-0.5.0.jar`
   * `tachyon-client-0.5.0.jar`

JoshRosen committed:
Disable MiMa checks for new unsafe module

SparkQA commented:
[Test build #31069 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31069/consoleFull) for   PR 5725 at commit [`0925847`](https://github.com/apache/spark/commit/092584701277394a704c7600c6a631326d7895c6).

SparkQA commented:
[Test build #31069 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31069/consoleFull) for   PR 5725 at commit [`0925847`](https://github.com/apache/spark/commit/092584701277394a704c7600c6a631326d7895c6).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public final class UnsafeFixedWidthAggregationMap `
  * `  public static class MapEntry `
  * `public final class UnsafeRow implements MutableRow `
  * `class UnsafeRowConverter(fieldTypes: Array[DataType]) `
  * `public final class PlatformDependent `
  * `public class ByteArrayMethods `
  * `public final class LongArray `
  * `public final class BitSet `
  * `public final class BitSetMethods `
  * `public final class Murmur3_x86_32 `
  * `public final class BytesToBytesMap `
  * `  public final class Location `
  * `  class Doubling implements HashMapGrowthStrategy `
  * `public class HeapMemoryAllocator implements MemoryAllocator `
  * `public class MemoryBlock extends MemoryLocation `
  * `public class MemoryLocation `
  * `public class UnsafeMemoryAllocator implements MemoryAllocator `

 * This patch **adds the following new dependencies:**
   * `spark-unsafe_2.10-1.4.0-SNAPSHOT.jar`

JoshRosen commented on unsafe/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java:
```diff
@@ -0,0 +1,618 @@
+      pageCursor += 8;
+      final long valueDataOffsetInPage = pageBaseOffset + pageCursor;
+      pageCursor += valueLengthBytes;
+      final long relativeOffsetFromKeyToValue = valueSizeOffsetInPage - keySizeOffsetInPage;
```
I think that this is redundant, actually, since the value always appears after the key, we know the key's length, and we always store the key and value in the same memory block.

If I remove this, we'll have an extra 32 bits of space in the long array that we can use for something else.  Moving to a 64-bit hashcode probably won't make a huge difference.  Is there another good use for this space?

JoshRosen committed:
Introduce MemoryManager interface; add to SparkEnv.

The configuration of HEAP vs UNSAFE is now done at the Spark core level.

The translation of encoded 64-bit addresses into base object + offset pairs
is now handled by MemoryManager, allowing this pointers to be safely passed
between operators that exchange data pages.

SparkQA commented:
[Test build #31095 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31095/consoleFull) for   PR 5725 at commit [`a8e4a3f`](https://github.com/apache/spark/commit/a8e4a3fe40574c3a609beeb4794b11bd720a31e7).

JoshRosen committed:
Don't redundantly store the offset from key to value, since we can compute this from the key size.

SparkQA commented:
[Test build #31101 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31101/consoleFull) for   PR 5725 at commit [`b45f070`](https://github.com/apache/spark/commit/b45f0708733e5777fff01c69206d0f1b8efcb7e9).

SparkQA commented:
[Test build #31095 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31095/consoleFull) for   PR 5725 at commit [`a8e4a3f`](https://github.com/apache/spark/commit/a8e4a3fe40574c3a609beeb4794b11bd720a31e7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public final class UnsafeFixedWidthAggregationMap `
  * `  public static class MapEntry `
  * `public final class UnsafeRow implements MutableRow `
  * `class UnsafeRowConverter(fieldTypes: Array[DataType]) `
  * `public final class PlatformDependent `
  * `public class ByteArrayMethods `
  * `public final class LongArray `
  * `public final class BitSet `
  * `public final class BitSetMethods `
  * `public final class Murmur3_x86_32 `
  * `public final class BytesToBytesMap `
  * `  public final class Location `
  * `  class Doubling implements HashMapGrowthStrategy `
  * `public class HeapMemoryAllocator implements MemoryAllocator `
  * `public class MemoryBlock extends MemoryLocation `
  * `public class MemoryLocation `
  * `public final class MemoryManager `
  * `public class UnsafeMemoryAllocator implements MemoryAllocator `

 * This patch **removes the following dependencies:**
   * `RoaringBitmap-0.4.5.jar`
   * `activation-1.1.jar`
   * `akka-actor_2.10-2.3.4-spark.jar`
   * `akka-remote_2.10-2.3.4-spark.jar`
   * `akka-slf4j_2.10-2.3.4-spark.jar`
   * `aopalliance-1.0.jar`
   * `arpack_combined_all-0.1.jar`
   * `avro-1.7.7.jar`
   * `breeze-macros_2.10-0.11.2.jar`
   * `breeze_2.10-0.11.2.jar`
   * `chill-java-0.5.0.jar`
   * `chill_2.10-0.5.0.jar`
   * `commons-beanutils-1.7.0.jar`
   * `commons-beanutils-core-1.8.0.jar`
   * `commons-cli-1.2.jar`
   * `commons-codec-1.10.jar`
   * `commons-collections-3.2.1.jar`
   * `commons-compress-1.4.1.jar`
   * `commons-configuration-1.6.jar`
   * `commons-digester-1.8.jar`
   * `commons-httpclient-3.1.jar`
   * `commons-io-2.1.jar`
   * `commons-lang-2.5.jar`
   * `commons-lang3-3.3.2.jar`
   * `commons-math-2.1.jar`
   * `commons-math3-3.4.1.jar`
   * `commons-net-2.2.jar`
   * `compress-lzf-1.0.0.jar`
   * `config-1.2.1.jar`
   * `core-1.1.2.jar`
   * `curator-client-2.4.0.jar`
   * `curator-framework-2.4.0.jar`
   * `curator-recipes-2.4.0.jar`
   * `gmbal-api-only-3.0.0-b023.jar`
   * `grizzly-framework-2.1.2.jar`
   * `grizzly-http-2.1.2.jar`
   * `grizzly-http-server-2.1.2.jar`
   * `grizzly-http-servlet-2.1.2.jar`
   * `grizzly-rcm-2.1.2.jar`
   * `groovy-all-2.3.7.jar`
   * `guava-14.0.1.jar`
   * `guice-3.0.jar`
   * `hadoop-annotations-2.2.0.jar`
   * `hadoop-auth-2.2.0.jar`
   * `hadoop-client-2.2.0.jar`
   * `hadoop-common-2.2.0.jar`
   * `hadoop-hdfs-2.2.0.jar`
   * `hadoop-mapreduce-client-app-2.2.0.jar`
   * `hadoop-mapreduce-client-common-2.2.0.jar`
   * `hadoop-mapreduce-client-core-2.2.0.jar`
   * `hadoop-mapreduce-client-jobclient-2.2.0.jar`
   * `hadoop-mapreduce-client-shuffle-2.2.0.jar`
   * `hadoop-yarn-api-2.2.0.jar`
   * `hadoop-yarn-client-2.2.0.jar`
   * `hadoop-yarn-common-2.2.0.jar`
   * `hadoop-yarn-server-common-2.2.0.jar`
   * `ivy-2.4.0.jar`
   * `jackson-annotations-2.4.0.jar`
   * `jackson-core-2.4.4.jar`
   * `jackson-core-asl-1.8.8.jar`
   * `jackson-databind-2.4.4.jar`
   * `jackson-jaxrs-1.8.8.jar`
   * `jackson-mapper-asl-1.8.8.jar`
   * `jackson-module-scala_2.10-2.4.4.jar`
   * `jackson-xc-1.8.8.jar`
   * `jansi-1.4.jar`
   * `javax.inject-1.jar`
   * `javax.servlet-3.0.0.v201112011016.jar`
   * `javax.servlet-3.1.jar`
   * `javax.servlet-api-3.0.1.jar`
   * `jaxb-api-2.2.2.jar`
   * `jaxb-impl-2.2.3-1.jar`
   * `jcl-over-slf4j-1.7.10.jar`
   * `jersey-client-1.9.jar`
   * `jersey-core-1.9.jar`
   * `jersey-grizzly2-1.9.jar`
   * `jersey-guice-1.9.jar`
   * `jersey-json-1.9.jar`
   * `jersey-server-1.9.jar`
   * `jersey-test-framework-core-1.9.jar`
   * `jersey-test-framework-grizzly2-1.9.jar`
   * `jets3t-0.7.1.jar`
   * `jettison-1.1.jar`
   * `jetty-util-6.1.26.jar`
   * `jline-0.9.94.jar`
   * `jline-2.10.4.jar`
   * `jodd-core-3.6.3.jar`
   * `json4s-ast_2.10-3.2.10.jar`
   * `json4s-core_2.10-3.2.10.jar`
   * `json4s-jackson_2.10-3.2.10.jar`
   * `jsr305-1.3.9.jar`
   * `jtransforms-2.4.0.jar`
   * `jul-to-slf4j-1.7.10.jar`
   * `kryo-2.21.jar`
   * `log4j-1.2.17.jar`
   * `lz4-1.2.0.jar`
   * `management-api-3.0.0-b012.jar`
   * `mesos-0.21.0-shaded-protobuf.jar`
   * `metrics-core-3.1.0.jar`
   * `metrics-graphite-3.1.0.jar`
   * `metrics-json-3.1.0.jar`
   * `metrics-jvm-3.1.0.jar`
   * `minlog-1.2.jar`
   * `netty-3.8.0.Final.jar`
   * `netty-all-4.0.23.Final.jar`
   * `objenesis-1.2.jar`
   * `opencsv-2.3.jar`
   * `oro-2.0.8.jar`
   * `paranamer-2.6.jar`
   * `parquet-column-1.6.0rc3.jar`
   * `parquet-common-1.6.0rc3.jar`
   * `parquet-encoding-1.6.0rc3.jar`
   * `parquet-format-2.2.0-rc1.jar`
   * `parquet-generator-1.6.0rc3.jar`
   * `parquet-hadoop-1.6.0rc3.jar`
   * `parquet-jackson-1.6.0rc3.jar`
   * `protobuf-java-2.4.1.jar`
   * `protobuf-java-2.5.0-spark.jar`
   * `py4j-0.8.2.1.jar`
   * `pyrolite-2.0.1.jar`
   * `quasiquotes_2.10-2.0.1.jar`
   * `reflectasm-1.07-shaded.jar`
   * `scala-compiler-2.10.4.jar`
   * `scala-library-2.10.4.jar`
   * `scala-reflect-2.10.4.jar`
   * `scalap-2.10.4.jar`
   * `scalatest_2.10-2.2.1.jar`
   * `slf4j-api-1.7.10.jar`
   * `slf4j-log4j12-1.7.10.jar`
   * `snappy-java-1.1.1.7.jar`
   * `spark-bagel_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-catalyst_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-core_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-graphx_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-launcher_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-mllib_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-network-common_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-network-shuffle_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-repl_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-sql_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-streaming_2.10-1.4.0-SNAPSHOT.jar`
   * `spire-macros_2.10-0.7.4.jar`
   * `spire_2.10-0.7.4.jar`
   * `stax-api-1.0.1.jar`
   * `stream-2.7.0.jar`
   * `tachyon-0.6.4.jar`
   * `tachyon-client-0.6.4.jar`
   * `uncommons-maths-1.2.2a.jar`
   * `unused-1.0.0.jar`
   * `xmlenc-0.52.jar`
   * `xz-1.0.jar`
   * `zookeeper-3.4.5.jar`

SparkQA commented:
[Test build #31101 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31101/consoleFull) for   PR 5725 at commit [`b45f070`](https://github.com/apache/spark/commit/b45f0708733e5777fff01c69206d0f1b8efcb7e9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public final class UnsafeFixedWidthAggregationMap `
  * `  public static class MapEntry `
  * `public final class UnsafeRow implements MutableRow `
  * `class UnsafeRowConverter(fieldTypes: Array[DataType]) `
  * `public final class PlatformDependent `
  * `public class ByteArrayMethods `
  * `public final class LongArray `
  * `public final class BitSet `
  * `public final class BitSetMethods `
  * `public final class Murmur3_x86_32 `
  * `public final class BytesToBytesMap `
  * `  public final class Location `
  * `  class Doubling implements HashMapGrowthStrategy `
  * `public class HeapMemoryAllocator implements MemoryAllocator `
  * `public class MemoryBlock extends MemoryLocation `
  * `public class MemoryLocation `
  * `public final class MemoryManager `
  * `public class UnsafeMemoryAllocator implements MemoryAllocator `

 * This patch **removes the following dependencies:**
   * `RoaringBitmap-0.4.5.jar`
   * `activation-1.1.jar`
   * `akka-actor_2.10-2.3.4-spark.jar`
   * `akka-remote_2.10-2.3.4-spark.jar`
   * `akka-slf4j_2.10-2.3.4-spark.jar`
   * `aopalliance-1.0.jar`
   * `arpack_combined_all-0.1.jar`
   * `avro-1.7.7.jar`
   * `breeze-macros_2.10-0.11.2.jar`
   * `breeze_2.10-0.11.2.jar`
   * `chill-java-0.5.0.jar`
   * `chill_2.10-0.5.0.jar`
   * `commons-beanutils-1.7.0.jar`
   * `commons-beanutils-core-1.8.0.jar`
   * `commons-cli-1.2.jar`
   * `commons-codec-1.10.jar`
   * `commons-collections-3.2.1.jar`
   * `commons-compress-1.4.1.jar`
   * `commons-configuration-1.6.jar`
   * `commons-digester-1.8.jar`
   * `commons-httpclient-3.1.jar`
   * `commons-io-2.1.jar`
   * `commons-lang-2.5.jar`
   * `commons-lang3-3.3.2.jar`
   * `commons-math-2.1.jar`
   * `commons-math3-3.4.1.jar`
   * `commons-net-2.2.jar`
   * `compress-lzf-1.0.0.jar`
   * `config-1.2.1.jar`
   * `core-1.1.2.jar`
   * `curator-client-2.4.0.jar`
   * `curator-framework-2.4.0.jar`
   * `curator-recipes-2.4.0.jar`
   * `gmbal-api-only-3.0.0-b023.jar`
   * `grizzly-framework-2.1.2.jar`
   * `grizzly-http-2.1.2.jar`
   * `grizzly-http-server-2.1.2.jar`
   * `grizzly-http-servlet-2.1.2.jar`
   * `grizzly-rcm-2.1.2.jar`
   * `groovy-all-2.3.7.jar`
   * `guava-14.0.1.jar`
   * `guice-3.0.jar`
   * `hadoop-annotations-2.2.0.jar`
   * `hadoop-auth-2.2.0.jar`
   * `hadoop-client-2.2.0.jar`
   * `hadoop-common-2.2.0.jar`
   * `hadoop-hdfs-2.2.0.jar`
   * `hadoop-mapreduce-client-app-2.2.0.jar`
   * `hadoop-mapreduce-client-common-2.2.0.jar`
   * `hadoop-mapreduce-client-core-2.2.0.jar`
   * `hadoop-mapreduce-client-jobclient-2.2.0.jar`
   * `hadoop-mapreduce-client-shuffle-2.2.0.jar`
   * `hadoop-yarn-api-2.2.0.jar`
   * `hadoop-yarn-client-2.2.0.jar`
   * `hadoop-yarn-common-2.2.0.jar`
   * `hadoop-yarn-server-common-2.2.0.jar`
   * `ivy-2.4.0.jar`
   * `jackson-annotations-2.4.0.jar`
   * `jackson-core-2.4.4.jar`
   * `jackson-core-asl-1.8.8.jar`
   * `jackson-databind-2.4.4.jar`
   * `jackson-jaxrs-1.8.8.jar`
   * `jackson-mapper-asl-1.8.8.jar`
   * `jackson-module-scala_2.10-2.4.4.jar`
   * `jackson-xc-1.8.8.jar`
   * `jansi-1.4.jar`
   * `javax.inject-1.jar`
   * `javax.servlet-3.0.0.v201112011016.jar`
   * `javax.servlet-3.1.jar`
   * `javax.servlet-api-3.0.1.jar`
   * `jaxb-api-2.2.2.jar`
   * `jaxb-impl-2.2.3-1.jar`
   * `jcl-over-slf4j-1.7.10.jar`
   * `jersey-client-1.9.jar`
   * `jersey-core-1.9.jar`
   * `jersey-grizzly2-1.9.jar`
   * `jersey-guice-1.9.jar`
   * `jersey-json-1.9.jar`
   * `jersey-server-1.9.jar`
   * `jersey-test-framework-core-1.9.jar`
   * `jersey-test-framework-grizzly2-1.9.jar`
   * `jets3t-0.7.1.jar`
   * `jettison-1.1.jar`
   * `jetty-util-6.1.26.jar`
   * `jline-0.9.94.jar`
   * `jline-2.10.4.jar`
   * `jodd-core-3.6.3.jar`
   * `json4s-ast_2.10-3.2.10.jar`
   * `json4s-core_2.10-3.2.10.jar`
   * `json4s-jackson_2.10-3.2.10.jar`
   * `jsr305-1.3.9.jar`
   * `jtransforms-2.4.0.jar`
   * `jul-to-slf4j-1.7.10.jar`
   * `kryo-2.21.jar`
   * `log4j-1.2.17.jar`
   * `lz4-1.2.0.jar`
   * `management-api-3.0.0-b012.jar`
   * `mesos-0.21.0-shaded-protobuf.jar`
   * `metrics-core-3.1.0.jar`
   * `metrics-graphite-3.1.0.jar`
   * `metrics-json-3.1.0.jar`
   * `metrics-jvm-3.1.0.jar`
   * `minlog-1.2.jar`
   * `netty-3.8.0.Final.jar`
   * `netty-all-4.0.23.Final.jar`
   * `objenesis-1.2.jar`
   * `opencsv-2.3.jar`
   * `oro-2.0.8.jar`
   * `paranamer-2.6.jar`
   * `parquet-column-1.6.0rc3.jar`
   * `parquet-common-1.6.0rc3.jar`
   * `parquet-encoding-1.6.0rc3.jar`
   * `parquet-format-2.2.0-rc1.jar`
   * `parquet-generator-1.6.0rc3.jar`
   * `parquet-hadoop-1.6.0rc3.jar`
   * `parquet-jackson-1.6.0rc3.jar`
   * `protobuf-java-2.4.1.jar`
   * `protobuf-java-2.5.0-spark.jar`
   * `py4j-0.8.2.1.jar`
   * `pyrolite-2.0.1.jar`
   * `quasiquotes_2.10-2.0.1.jar`
   * `reflectasm-1.07-shaded.jar`
   * `scala-compiler-2.10.4.jar`
   * `scala-library-2.10.4.jar`
   * `scala-reflect-2.10.4.jar`
   * `scalap-2.10.4.jar`
   * `scalatest_2.10-2.2.1.jar`
   * `slf4j-api-1.7.10.jar`
   * `slf4j-log4j12-1.7.10.jar`
   * `snappy-java-1.1.1.7.jar`
   * `spark-bagel_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-catalyst_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-core_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-graphx_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-launcher_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-mllib_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-network-common_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-network-shuffle_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-repl_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-sql_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-streaming_2.10-1.4.0-SNAPSHOT.jar`
   * `spire-macros_2.10-0.7.4.jar`
   * `spire_2.10-0.7.4.jar`
   * `stax-api-1.0.1.jar`
   * `stream-2.7.0.jar`
   * `tachyon-0.6.4.jar`
   * `tachyon-client-0.6.4.jar`
   * `uncommons-maths-1.2.2a.jar`
   * `unused-1.0.0.jar`
   * `xmlenc-0.52.jar`
   * `xz-1.0.jar`
   * `zookeeper-3.4.5.jar`

JoshRosen committed:
Fix test compilation

SparkQA commented:
[Test build #31113 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31113/consoleFull) for   PR 5725 at commit [`162caf7`](https://github.com/apache/spark/commit/162caf74c15952f6ae0482c1fa74529a0289b039).

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:
```diff
@@ -140,7 +140,9 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
-              planLater(child))) :: Nil
+              planLater(child),
+              unsafeEnabled),
+          unsafeEnabled) :: Nil
```
indent off here

rxin commented on unsafe/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java:
```diff
@@ -0,0 +1,552 @@
+ * This is backed by a power-of-2-sized hash table, using quadratic probing with triangular numbers,
+ * which is guaranteed to exhaust the space.
+ * <p>
+ * Note that even though we use long for indexing, the map can support up to 2^31 keys because
```
this is no longer true - we use 32-bit int for index now.

rxin commented on unsafe/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java:
should be nano sec

rxin commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java:
```diff
@@ -0,0 +1,251 @@
+    }
+    System.out.println("Average probes per lookup: " + map.getAverageProbesPerLookup());
+    System.out.println("Number of hash collisions: " + map.getNumHashCollisions());
+    System.out.println("Time spent resizing (ms): " + map.getTimeSpentResizingMs());
```
ms -> ns

rxin commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java:
```diff
@@ -0,0 +1,428 @@
+   * methods.
+   */
+  @Nullable
+  private StructType schema;
```
I think we can remove this -- since internally we never use this.

JoshRosen replied:
It's used in the generic `get()` call, which is only called internally by generated code that accesses UTF8String columns.  If we add an internal `getUTF8String()` method and update the code generator to use this, then we can completely remove this.  I might take a stab at this, since I think it's a pretty minimal change (I'll clearly mark the new method as internal so that it gets handled right during the planned InternalRow / Row split).

rxin commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java:
```diff
@@ -0,0 +1,251 @@
+    // Zero out the buffer that's used to hold the current row. This is necessary in order
+    // to ensure that rows hash properly, since garbage data from the previous row could
+    // otherwise end up as padding in this row.
+    Arrays.fill(groupingKeyConversionScratchSpace, 0);
```
technically we only need to zero out the beginning, no?

JoshRosen replied:
Yes.  I think that `Array.fill` is reasonably fast because it can be translated into SIMD operations, but it still doesn't hurt to zero fewer bytes.  This could actually matter in cases where we grow the size of `groupingKeyConversionScratchSpace` in responses to a single large key, since the current code could wind up zeroing much more space than is actually needed.  I'll update this.

rxin commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java:
```diff
@@ -0,0 +1,251 @@
+    return currentAggregationBuffer;
+  }
+
+  public static class MapEntry {
```
maybe add comment saying this is only used for iterator.

rxin commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java:
```diff
@@ -0,0 +1,428 @@
+
+  private Object baseObject;
+  private long baseOffset;
+  /** The number of fields in this row, used for calculating the bitset width (and in assertions) */
```
nitpick: if we have comments, let's add a blank line

rxin commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java:
```diff
@@ -0,0 +1,428 @@
+ * base address of the row) that points to the beginning of the variable-length field.
+ *
+ * Instances of `UnsafeRow` act as pointers to row data stored in this format, similar to how
+ * `Writable` objects work in Hadoop.
```
how is this similar to writable?

JoshRosen replied:
It's a very superficial similarity in the sense that both are mutable objects that are used to expose data that's backed by something that's serialized. This comment is confusing, though, so I'm going to remove it.

rxin commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java:
```diff
@@ -0,0 +1,428 @@
+  static {
+    settableFieldTypes = Collections.unmodifiableSet(
+      new HashSet<DataType>(
+        Arrays.asList(new DataType[] {
```
date and timestamp should also be supported?

we can do it in a future PR though -- maybe create jira tickets under https://issues.apache.org/jira/browse/SPARK-7075

JoshRosen replied:
Done: https://issues.apache.org/jira/browse/SPARK-7199

SparkQA commented:
[Test build #31113 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31113/consoleFull) for   PR 5725 at commit [`162caf7`](https://github.com/apache/spark/commit/162caf74c15952f6ae0482c1fa74529a0289b039).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public final class UnsafeFixedWidthAggregationMap `
  * `  public static class MapEntry `
  * `public final class UnsafeRow implements MutableRow `
  * `class UnsafeRowConverter(fieldTypes: Array[DataType]) `
  * `public final class PlatformDependent `
  * `public class ByteArrayMethods `
  * `public final class LongArray `
  * `public final class BitSet `
  * `public final class BitSetMethods `
  * `public final class Murmur3_x86_32 `
  * `public final class BytesToBytesMap `
  * `  public final class Location `
  * `  class Doubling implements HashMapGrowthStrategy `
  * `public class HeapMemoryAllocator implements MemoryAllocator `
  * `public class MemoryBlock extends MemoryLocation `
  * `public class MemoryLocation `
  * `public final class MemoryManager `
  * `public class UnsafeMemoryAllocator implements MemoryAllocator `

 * This patch **adds the following new dependencies:**
   * `spark-unsafe_2.10-1.4.0-SNAPSHOT.jar`

rxin commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java:
```diff
@@ -0,0 +1,428 @@
+    }
+  }
+
+  public UTF8String getUTF8String(int i) {
```
i created https://issues.apache.org/jira/browse/SPARK-7190 for utf8string backed by memory address

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/UnsafeRowConverter.scala:
```diff
@@ -0,0 +1,226 @@
+}
+
+private class IntUnsafeColumnWriter private() extends PrimitiveUnsafeColumnWriter[Int] {
+  override def write(
```
can u define the return value?

JoshRosen replied:
The return value is the number of bytes written, which we use to increment the position cursor.  I think that I commented on this in the base `UnsafeColumnWriter` class.

JoshRosen replied:
More specifically, this is the number of variable-length bytes written.  Primitive types, like int, don't write to the variable-length portion and thus return 0.

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/GeneratedAggregate.scala:
```diff
@@ -265,7 +284,49 @@ case class GeneratedAggregate(
+          aggregationBufferSchema,
+          groupKeySchema,
+          SparkEnv.get.unsafeMemoryManager,
+          1024 * 16,
```
use named argument for this

JoshRosen replied:
AFAIK we can't use the Scala named argument syntax because `UnsafeFixedWidthAggregationMap` is implemented in Java, but I can add a line comment explaining this.  Also, I wonder if there's a better choice of default map size here.  If we had statistics / estimates from the optimizer, we could make a better choice here.

rxin replied:
ah ok nvm. then let's add a comment like

```java
1024 * 16,  // page size
...
```

rxin commented on unsafe/src/main/java/org/apache/spark/unsafe/memory/MemoryBlock.java:
```diff
@@ -0,0 +1,74 @@
+  /**
+   * Creates a memory block pointing to the memory used by the byte array.
+   */
+  public static MemoryBlock fromByteArray(final byte[] array) {
```
can we remove this since it is not used?

JoshRosen replied:
It's called in a test method in `TestLongArray.scala`, but we don't need it.

rxin commented on unsafe/src/main/java/org/apache/spark/unsafe/memory/MemoryManager.java:
```diff
@@ -0,0 +1,176 @@
+ * <p>
+ * Instead, we use the following approach to encode record pointers in 64-bit longs: for off-heap
+ * mode, just store the raw address, and for on-heap mode use the upper 13 bits of the address to
+ * store a "page number" and the lower 51 bits to store an offset within this page. These page
```
we should spell out the total number of pages and memory addressable with this encoding

JoshRosen replied:
Done.

rxin commented on unsafe/src/main/java/org/apache/spark/unsafe/memory/MemoryManager.java:
```diff
@@ -0,0 +1,176 @@
+  /**
+   * The number of entries in the page table.
+   */
+  private static final int PAGE_TABLE_SIZE = (int) 1L << 13;
```
why not just ```1 << 13``` ?

JoshRosen replied:
Good catch; fixed.

rxin commented on unsafe/src/main/java/org/apache/spark/unsafe/memory/MemoryBlock.java:
```diff
@@ -0,0 +1,74 @@
+   * Optional page number; used when this MemoryBlock represents a page allocated by a
+   * MemoryManager. This is package-private and is modified by MemoryManager.
+   */
+  int pageNumber = -1;
```
would this be slightly faster if it is a long? since we do quite a bit of shifting and casting from int to long for this field

JoshRosen replied:
My original motivation for making this an int was the fact that our page table is indexed by an integer, but now that I look more closely it looks like it's most-often read by `encodePageNumberAndOffset`, which does the long cast.  If we wanted to go a bit further in terms of saving instructions here, I guess we could store the pre-shifted page number then do the reverse shift for the accesses in `freePage`, since that's performed far fewer times.

rxin commented on unsafe/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java:
call out explicitly the strength reduction here. Essentially a mod operation, but done using a bit mask since this is power-of-2 sized hash map.

rxin commented on unsafe/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java:
should say what the 8 + 8 are (key len + value len?)

rxin commented on unsafe/src/main/java/org/apache/spark/unsafe/memory/MemoryManager.java:
```diff
@@ -0,0 +1,176 @@
+   * Allocate a block of memory that will be tracked in the MemoryManager's page table; this is
+   * intended for allocating large blocks of memory that will be shared between operators.
+   */
+  public MemoryBlock allocatePage(long size) {
```
would be great to log something here

JoshRosen replied:
Agreed.  This package currently has no dependencies on `core`, so we can't use the `Logging` trait here.  Should we use the lower-level logging library directly, or just add the core dependency?

rxin replied:
We can use the lower level logging library directly, like what the network module does.

JoshRosen replied:
Done.

rxin commented on unsafe/src/main/java/org/apache/spark/unsafe/memory/MemoryManager.java:
```diff
@@ -0,0 +1,176 @@
+  /**
+   * Free a block of memory allocated via {@link MemoryManager#allocatePage(long)}.
+   */
+  public void freePage(MemoryBlock page) {
```
and log something here too

JoshRosen replied:
Done.

rxin commented:
I went through and left some comments. Kudos - this is super well done!

JoshRosen commented:
When running with off-heap, one concern is proper cleanup of managed memory when tasks fail or are cancelled.  Given that we currently do not share managed memory between tasks running on the same executor, I'm thinking that we should probably move MemoryManager from SparkEnv to TaskContext.  This would let us perform leak-detection when tasks complete; it also simplifies concurrency issues (no contention for the MemoryManager's lock when allocating / freeing blocks, provided that tasks are single-threaded) and reduces the chance that we'll exhaust our page table limit, since each task will have its own independent table.

JoshRosen commented on unsafe/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java:
Yep, that's key length and value length.  I use the same pattern in a couple of other places, so maybe I can pull this into a constant or add comments there, too.

rxin commented:
I think good memory allocator implementations usually have a thread local cache to reduce contention. We can do the same thing too, can't we? The problem with only allocating at the TaskContext level is that you then lack global coordination, and as a result, each task is bounded by total_mem/num_tasks memory, rather than a more dynamic allocation.

Maybe a good way to do this is to have the global memory allocator in SparkEnv, and TaskContext tracks the individual pages used by each task, so it can free it.

rxin commented:
update: thought about it more. I don't think we need to worry about contention at the moment, since our allocation unit is pretty large.

JoshRosen committed:
Only zero the used portion of groupingKeyConversionScratchSpace

JoshRosen committed:
Measure timeSpentResizing in nanoseconds instead of milliseconds.

JoshRosen committed:
More comments, formatting, and code cleanup.

JoshRosen committed:
Add logging to MemoryManager

JoshRosen committed:
Rename unsafe Java test suites to match Scala test naming convention.

This is necessary in order for SBT to recognize and run these suites.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31173 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31173/consoleFull) for   PR 5725 at commit [`a19e066`](https://github.com/apache/spark/commit/a19e0661ac9338d2856883537e314296607901a4).

JoshRosen committed:
Fix debug vs. trace in logging message.

JoshRosen committed:
Remove an unused method from ByteArrayMethods.

JoshRosen committed:
Split MemoryManager into ExecutorMemoryManager and TaskMemoryManager:

- Implement memory leak detection, with exception vs. logging controlled by
  a configuration option.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

JoshRosen commented on unsafe/src/main/java/org/apache/spark/unsafe/memory/TaskMemoryManager.java:
```diff
@@ -0,0 +1,237 @@
+   * Tracks memory allocated with {@link TaskMemoryManager#allocate(long)}, used to detect / clean
+   * up leaked memory.
+   */
+  private final HashSet<MemoryBlock> allocatedNonPageMemory = new HashSet<MemoryBlock>();
```
Here, I've used a HashSet to track non-page allocations.  I don't think that this should be _too_ expensive because we only use these in a couple of places (long arrays and bitsets, which we might just want to always keep on-heap anyways (in which case we don't need to involve the allocator)).  I didn't implement equals() or hashCode() for MemoryBlock, but I think that should be okay; object identity is fine for our purposes here.

SparkQA commented:
[Test build #31185 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31185/consoleFull) for   PR 5725 at commit [`70a39e4`](https://github.com/apache/spark/commit/70a39e4fe05c31ac73b99614115f1fe9f150e1e6).

SparkQA commented:
**[Test build #31173 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31173/consoleFull)**     for PR 5725 at commit [`a19e066`](https://github.com/apache/spark/commit/a19e0661ac9338d2856883537e314296607901a4)     after a configured wait of `150m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31173/
Test FAILed.

rxin commented on core/src/main/scala/org/apache/spark/executor/Executor.scala:
```diff
@@ -207,7 +210,23 @@ private[spark] class Executor(
+        } finally {
+          // Release managed memory used by this task
+          val freedMemory = taskMemoryManager.cleanUpAllAllocatedMemory()
+          if (succeeded && freedMemory > 0) {
```
maybe we want to enforce so that even in the case of failures, memory should be cleaned up?

also would be great to not have two different code path in executor vs dagscheduler

JoshRosen replied:
I originally considered putting this in Task, but unfortunately `runLocally` doesn't go through `Task`.    We could try to consolidate both of the code paths by using Task in `runLocally`, but I think this will be a larger change (we'd need changes higher up the DAGScheduler call stack to create tasks of the appropriate types, etc., and this might be hard to review).

As a compromise, I'm going to leave the duplicated code in place but add comments so that readers are aware of the duplication and remember to change both copies when making changes.

JoshRosen replied:
I also agree that we should enforce this for failures as well.  In the worst case, I suppose our failure message is masked by a memory leak message, but that's fine for now as this is disabled by default and is only used in tests.

marmbrus commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/UnsafeRowConverter.scala:
```diff
@@ -0,0 +1,226 @@
+        unsafeRow.setNullAt(fieldNumber)
+      } else {
+        appendCursor += writers(fieldNumber).write(
+          row(fieldNumber),
```
This is pretty minor, but is there a reason we are boxing here instead of passing the row itself in, allowing a specific accessor to be used for extraction?

JoshRosen replied:
That's a really good idea; can't believe I didn't think of that...

Should be an easy change, so I'll do this shortly.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/GeneratedAggregate.scala:
```diff
@@ -265,7 +283,49 @@ case class GeneratedAggregate(
+          groupKeySchema,
+          TaskContext.get.taskMemoryManager(),
+          1024 * 16, // initial capacity
+          false // disable tracking of performance metrics
```
nit: consider named arguments instead of comments.

JoshRosen replied:
Reynold made the same comment upthread; we can't use Scala-style named arguments here because `UnsafFixedWidthAggregationMap` is a Java class.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/GeneratedAggregate.scala:
```diff
@@ -41,13 +42,15 @@ case class AggregateEvaluation(
     aggregateExpressions: Seq[NamedExpression],
-    child: SparkPlan)
+    child: SparkPlan,
+    unsafeEnabled: Boolean)
```
nit: typically place children last (although we aren't great about following this convention)...

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/GeneratedAggregate.scala:
```diff
@@ -41,13 +42,15 @@ case class AggregateEvaluation(
+ * @param unsafeEnabled whether to allow Unsafe-based aggregation buffers to be used.
  */
 @DeveloperApi
 case class GeneratedAggregate(
```
We should consider breaking this into two operators at some point so the explain is more readable.

JoshRosen replied:
Agreed.  I'd like to avoid duplication of the code generation logic, though.

JoshRosen committed:
Throw memory leak warning even in case of error; add warning about code duplication

SparkQA commented:
[Test build #31185 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31185/consoleFull) for   PR 5725 at commit [`70a39e4`](https://github.com/apache/spark/commit/70a39e4fe05c31ac73b99614115f1fe9f150e1e6).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public final class UnsafeFixedWidthAggregationMap `
  * `  public static class MapEntry `
  * `public final class UnsafeRow implements MutableRow `
  * `class UnsafeRowConverter(fieldTypes: Array[DataType]) `
  * `public final class PlatformDependent `
  * `public class ByteArrayMethods `
  * `public final class LongArray `
  * `public final class BitSet `
  * `public final class BitSetMethods `
  * `public final class Murmur3_x86_32 `
  * `public final class BytesToBytesMap `
  * `  public final class Location `
  * `  class Doubling implements HashMapGrowthStrategy `
  * `public class ExecutorMemoryManager `
  * `public class HeapMemoryAllocator implements MemoryAllocator `
  * `public class MemoryBlock extends MemoryLocation `
  * `public class MemoryLocation `
  * `public final class TaskMemoryManager `
  * `public class UnsafeMemoryAllocator implements MemoryAllocator `

 * This patch **adds the following new dependencies:**
   * `spark-unsafe_2.10-1.4.0-SNAPSHOT.jar`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31185/
Test FAILed.

JoshRosen commented:
The most recent test failure looks like a possible double-free somewhere.  Here's what I saw when I ran that test locally:

```
[info] Test org.apache.spark.unsafe.map.BytesToBytesMapOffHeapSuite.randomizedStressTest started
java(7099,0x111e21000) malloc: *** error for object 0x7fafec90e800: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
```

This error is non-deterministic.  Based on some local debugging with `Thread.dumpStack()`, I think that this was due to a race between BytesToBytesMap's finalizer and the tearDown() method in AbstractBytesToBytesMapSuite.  I think that we should remove the finalizers, since they add complexity and are no longer necessary since we have leak detection / cleanup at higher levels.

JoshRosen committed:
Remove BytesToBytesMap.finalize()

This is no longer necessary now that we do leak
detection and cleanup at higher levels (as part of MemoryManager).

JoshRosen commented:
As part of UnsafeRowConverter refactoring, I realized that I forgot to add support for a whole bunch of primitives, including booleans.  These are trivial to add, so I'll go ahead and do that now.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/GeneratedAggregate.scala:
```diff
@@ -265,7 +283,49 @@ case class GeneratedAggregate(
+              // though, we need to make a defensive copy of the result so that we don't return an
+              // object that might contain dangling pointers to the freed memory
+              val resultCopy = result.copy()
+              aggregationMap.free()
```
Checking my understanding here... We are safe for takes that don't exhaust the iterator because of  some registration that is happening due to `TaskContext.get.taskMemoryManager()` above?

JoshRosen replied:
We're safe in the sense that the memory will eventually get cleaned up, but I think that we would get memory leak exceptions if those warnings / checks have been enabled (the cleanup is always performed but whether we fail tasks is controlled by a flag).  To avoid this issue, I think we need to return a closable iterator interface which lets downstream consumers signal that they will read no more output.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/GeneratedAggregate.scala:
```diff
@@ -265,7 +283,49 @@ case class GeneratedAggregate(
 
         val resultProjection = resultProjectionBuilder()
         Iterator(resultProjection(buffer))
+      } else if (unsafeEnabled && schemaSupportsUnsafe) {
```
Similar to my comment above, I think it would be better if this decision was made in the query planner.

JoshRosen committed:
Refactor UnsafeRowConverter to avoid unnecessary boxing.

We now pass the source row into the method, allowing the converter to use type
specific accessors to extract column values.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala:
```diff
@@ -30,6 +30,7 @@ private[spark] object SQLConf {
   val DEFAULT_SIZE_IN_BYTES = "spark.sql.defaultSizeInBytes"
   val SHUFFLE_PARTITIONS = "spark.sql.shuffle.partitions"
   val CODEGEN_ENABLED = "spark.sql.codegen"
+  val UNSAFE_ENABLED = "spark.sql.unsafe.enabled"
```
unless we think there are going to be other options under unsafe,  we might consider getting rid of `enabled` to match `codegen`

JoshRosen replied:
There was an option for choosing heap vs. offheap, but that option got moved into Spark core.  I think we use the `enabled` convention a bit more in core.  I don't have strong feelings on this one.

JoshRosen committed:
Follow 'place children last' convention for GeneratedAggregate

JoshRosen committed:
Add converters for Null, Boolean, Byte, and Short columns.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31202 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31202/consoleFull) for   PR 5725 at commit [`eeee512`](https://github.com/apache/spark/commit/eeee512bd94d463f741170e904ae186e238f997c).

rxin commented:
LGTM. There are some issues, but we can easily address them in followup PRs.

SparkQA commented:
[Test build #31202 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31202/consoleFull) for   PR 5725 at commit [`eeee512`](https://github.com/apache/spark/commit/eeee512bd94d463f741170e904ae186e238f997c).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public final class UnsafeFixedWidthAggregationMap `
  * `  public static class MapEntry `
  * `public final class UnsafeRow implements MutableRow `
  * `class UnsafeRowConverter(fieldTypes: Array[DataType]) `
  * `public final class PlatformDependent `
  * `public class ByteArrayMethods `
  * `public final class LongArray `
  * `public final class BitSet `
  * `public final class BitSetMethods `
  * `public final class Murmur3_x86_32 `
  * `public final class BytesToBytesMap `
  * `  public final class Location `
  * `  class Doubling implements HashMapGrowthStrategy `
  * `public class ExecutorMemoryManager `
  * `public class HeapMemoryAllocator implements MemoryAllocator `
  * `public class MemoryBlock extends MemoryLocation `
  * `public class MemoryLocation `
  * `public final class TaskMemoryManager `
  * `public class UnsafeMemoryAllocator implements MemoryAllocator `

 * This patch **adds the following new dependencies:**
   * `spark-unsafe_2.10-1.4.0-SNAPSHOT.jar`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31202/
Test FAILed.

JoshRosen commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31245 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31245/consoleFull) for   PR 5725 at commit [`eeee512`](https://github.com/apache/spark/commit/eeee512bd94d463f741170e904ae186e238f997c).

SparkQA commented:
[Test build #732 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/732/consoleFull) for   PR 5725 at commit [`eeee512`](https://github.com/apache/spark/commit/eeee512bd94d463f741170e904ae186e238f997c).

SparkQA commented:
[Test build #31245 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31245/consoleFull) for   PR 5725 at commit [`eeee512`](https://github.com/apache/spark/commit/eeee512bd94d463f741170e904ae186e238f997c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public final class UnsafeFixedWidthAggregationMap `
  * `  public static class MapEntry `
  * `public final class UnsafeRow implements MutableRow `
  * `class UnsafeRowConverter(fieldTypes: Array[DataType]) `
  * `public final class PlatformDependent `
  * `public class ByteArrayMethods `
  * `public final class LongArray `
  * `public final class BitSet `
  * `public final class BitSetMethods `
  * `public final class Murmur3_x86_32 `
  * `public final class BytesToBytesMap `
  * `  public final class Location `
  * `  class Doubling implements HashMapGrowthStrategy `
  * `public class ExecutorMemoryManager `
  * `public class HeapMemoryAllocator implements MemoryAllocator `
  * `public class MemoryBlock extends MemoryLocation `
  * `public class MemoryLocation `
  * `public final class TaskMemoryManager `
  * `public class UnsafeMemoryAllocator implements MemoryAllocator `

 * This patch **adds the following new dependencies:**
   * `spark-unsafe_2.10-1.4.0-SNAPSHOT.jar`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31245/
Test PASSed.

rxin commented:
Ok merging in master. Let's make sure we address @marmbrus 's comments in followup prs.

tedyu commented:
<sup>**[sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java, line 133 \[r14\]](https://reviewable.io:443/reviews/apache/spark/5725#-Jo56lmUPtN4oJEAxApu)** ([raw file](https://github.com/apache/spark/blob/eeee512bd94d463f741170e904ae186e238f997c/sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java#L133)):</sup>
Please include writtenLength, unsafeRow.length and javaRow in the assertion so that we have more information when debugging.

---

<sup>**[sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java, line 148 \[r14\]](https://reviewable.io:443/reviews/apache/spark/5725#-Jo57pOAjJQa1nPxxd2y)** ([raw file](https://github.com/apache/spark/blob/eeee512bd94d463f741170e904ae186e238f997c/sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java#L148)):</sup>
Condition doesn't match assertion message w.r.t. <=

---

<sup>**[unsafe/src/main/java/org/apache/spark/unsafe/array/ByteArrayMethods.java, line 23 \[r14\]](https://reviewable.io:443/reviews/apache/spark/5725#-Jo5AmeniFC0_bDQdiyb)** ([raw file](https://github.com/apache/spark/blob/eeee512bd94d463f741170e904ae186e238f997c/unsafe/src/main/java/org/apache/spark/unsafe/array/ByteArrayMethods.java#L23)):</sup>
Declare this class static ?

---

<sup>**[unsafe/src/main/java/org/apache/spark/unsafe/array/LongArray.java, line 26 \[r14\]](https://reviewable.io:443/reviews/apache/spark/5725#-Jo5B0VcBmjKX4TnrZD7)** ([raw file](https://github.com/apache/spark/blob/eeee512bd94d463f741170e904ae186e238f997c/unsafe/src/main/java/org/apache/spark/unsafe/array/LongArray.java#L26)):</sup>
nit: in-heap -> on-heap

---

<sup>**[unsafe/src/main/java/org/apache/spark/unsafe/bitset/BitSet.java, line 24 \[r14\]](https://reviewable.io:443/reviews/apache/spark/5725#-Jo5BT5TTxBoPNV_juFI)** ([raw file](https://github.com/apache/spark/blob/eeee512bd94d463f741170e904ae186e238f997c/unsafe/src/main/java/org/apache/spark/unsafe/bitset/BitSet.java#L24)):</sup>
https://docs.oracle.com/javase/7/docs/api/java/util/BitSet.html doesn't serve your needs ?

---

<sup>**[unsafe/src/main/java/org/apache/spark/unsafe/bitset/BitSetMethods.java, line 75 \[r14\]](https://reviewable.io:443/reviews/apache/spark/5725#-Jo5C4Ty7HYh-hvVPSEb)** ([raw file](https://github.com/apache/spark/blob/eeee512bd94d463f741170e904ae186e238f997c/unsafe/src/main/java/org/apache/spark/unsafe/bitset/BitSetMethods.java#L75)):</sup>
Maybe use something similar to http://www.docjar.com/docs/api/sun/misc/Unsafe.html#getInt(Object, long) for speedup ?

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/apache/spark/5725)
<!-- Sent from Reviewable.io -->

tgravescs commented:
I'm not able to build after this went  in.  Any special requirements to build this?

[info] Compiling 15 Java sources to /home/tgraves/tgravescs_spark/unsafe/target/scala-2.10/classes...
[error] /home/tgraves/tgravescs_spark/unsafe/src/main/java/org/apache/spark/unsafe/PlatformDependent.java:22: warning: sun.misc.Unsafe is Sun proprietary API and may be removed in a future release
[error] import sun.misc.Unsafe;
[error]                ^
[error] /home/tgraves/tgravescs_spark/unsafe/src/main/java/org/apache/spark/unsafe/PlatformDependent.java:26: warning: sun.misc.Unsafe is Sun proprietary API and may be removed in a future release
[error]   public static final Unsafe UNSAFE;
[error]                       ^
[error] /home/tgraves/tgravescs_spark/unsafe/src/main/java/org/apache/spark/unsafe/PlatformDependent.java:43: warning: sun.misc.Unsafe is Sun proprietary API and may be removed in a future release
[error]     sun.misc.Unsafe unsafe;
[error]             ^
[error] /home/tgraves/tgravescs_spark/unsafe/src/main/java/org/apache/spark/unsafe/PlatformDependent.java:45: warning: sun.misc.Unsafe is Sun proprietary API and may be removed in a future release
[error]       Field unsafeField = Unsafe.class.getDeclaredField("theUnsafe");
[error]                           ^
[error] /home/tgraves/tgravescs_spark/unsafe/src/main/java/org/apache/spark/unsafe/PlatformDependent.java:47: warning: sun.misc.Unsafe is Sun proprietary API and may be removed in a future release
[error]       unsafe = (sun.misc.Unsafe) unsafeField.get(null);
[error]                         ^
[error] /home/tgraves/tgravescs_spark/unsafe/src/main/java/org/apache/spark/unsafe/memory/MemoryBlock.java:53: setMemory(long,long,byte) in sun.misc.Unsafe cannot be applied to (java.lang.Object,long,long,byte)
[error]     PlatformDependent.UNSAFE.setMemory(obj, offset, length, (byte) 0);
[error]                             ^
[error] 1 error
[error] 5 warnings
[error] Compile failed at Apr 30, 2015 2:52:04 PM [0.862s]

rxin commented:
Did you have warning turned on into error?

tgravescs commented:
not sure what your question is exactly, I think the last one is an error instead of warning.  I didn't specifying anything special in build command: mvn -Pyarn -Phadoop-2.4 -DskipTests clean package install

JoshRosen commented:
I'm looking into this issue now; these warnings can't be suppressed via the usual annotations mechanism; it requires a special compiler flag.  Planning to submit a patch soon; would be great to get your help in testing it.

tgravescs commented:
Thanks, Also I should have specified, I'm running on a rhel6 box using jdlk6

vanzin commented:
Java 6's output:

    $ javap sun.misc.Unsafe | grep setMemory
        public native void setMemory(long, long, byte);

So Spark is using an API that doesn't exist there.

vanzin commented:
On a side note, if Spark supports java 6, why are all jenkins builds running java 7?

JoshRosen commented:
@tgravescs I've opened #5814 to try to fix the compiler warnings (and error due to `Unsafe.setMemory`); it would be great if you could test out that commit and see whether it fixes things in your environment.

JoshRosen commented:
@vanzin There's an old JIRA for building w/ Java 6 on Jenkins; I've re-opened it: https://issues.apache.org/jira/browse/SPARK-1437

cloud-fan commented on unsafe/src/main/java/org/apache/spark/unsafe/memory/TaskMemoryManager.java:
```diff
@@ -0,0 +1,237 @@
+ * retrieve the base object.
+ * <p>
+ * This allows us to address 8192 pages. In on-heap mode, the maximum page size is limited by the
+ * maximum size of a long[] array, allowing us to address 8192 * 2^32 * 8 bytes, which is
```
Should this be `8192 * 2^31 * 8 bytes`? The length of array is up to `Integer.MAX_VALUE`.
